### PR TITLE
Configurable cache file name

### PR DESCRIPTION
**Pull request by apeneve**
Background for pull request: I have two nuget server sites sharing the same package repository folder. One public with authentication, and one internal. In this scenario I get file lock errors on the cache file, so I need to be able to configure different cache files for each site.

### CODE REVIEW

**Comment by joelverhagen** [comment]
@apeneve, sorry about the delay. Thanks for the contribution. I'm taking a look at this now!

File: src/NuGet.Server/Web.config [hunk 1]
```diff
@@ -22,6 +22,11 @@
     -->
     <add key="packagesPath" value="" />
 
+    <!--
+    Change the name of the internal cache file. Default is machine name (System.Environment.MachineName).
```
Comment (joelverhagen) [inline src/NuGet.Server/Web.config]: It's unclear whether putting a path here works (or even _should_ work).

I would recommend updated the XML comment to state that this must be a file name and not a relative or absolute path. Additionally, I would validate at runtime that this is just a file name. In the future we could extend this to support relative paths but that seems unnecessary now and can be expanded without being a breaking change.

File: src/NuGet.Server.Core/Infrastructure/ServerPackageRepository.cs [hunk 2]
```diff
@@ -105,9 +105,28 @@ public class ServerPackageRepository
...
+            const string suffix = ".cache.bin";
+
+            if (String.IsNullOrWhiteSpace(fileName))
+            {
+                // Default file name
+                return Environment.MachineName.ToLowerInvariant() + suffix;
+            }
+
+            if (fileName.EndsWith(suffix, StringComparison.OrdinalIgnoreCase))
+                return fileName;
```
Comment (joelverhagen) [inline src/NuGet.Server.Core/Infrastructure/ServerPackageRepository.cs]: nitpick: curly `{` `}` around all if blocks, even if one line.

File: test/NuGet.Server.Core.Tests/Infrastructure/FuncSettingsProvider.cs [hunk 3]
```diff
@@ -7,8 +7,8 @@ namespace NuGet.Server.Core.Tests.Infrastructure
 {
     class FuncSettingsProvider : ISettingsProvider
     {
-        readonly Func<string, bool, bool> _getSetting;
-        internal FuncSettingsProvider(Func<string,bool,bool> getSetting)
+        readonly Func<string, object, object> _getSetting;
+        internal FuncSettingsProvider(Func<string,object,object> getSetting)
```
Comment (joelverhagen) [inline test/NuGet.Server.Core.Tests/Infrastructure/FuncSettingsProvider.cs]: nitpic: spaces between type parameters after the comma `,`

**Comment by joelverhagen** [comment]
Could you add some unit or integration test that verified this behavior? I'm thinking something that verifies that if the this setting is provided, the cache file appears in the packages directory (`File.Exists` or something).

File: src/NuGet.Server/Web.config (hunk 1 above)
Comment (apeneve) [reply src/NuGet.Server/Web.config]: Agreed. I updated the XML comment and added validation of a valid file name.

**Comment by apeneve** [comment]
@joelverhagen  I fixed up the nitpic comments and extended the XML comment a bit. I also added some tests to the ServerPackageRepository class to describe the different the cache file name scenarios.

**Comment by joelverhagen** [comment]
@apeneve, wonderful! I'll work on merging this and shipping this as a `3.1.0` package.

**Comment by joelverhagen** [comment]
@apeneve, this is available as 3.1.0 on nuget.org:
https://www.nuget.org/packages/NuGet.Server/3.1.0

**Comment by apeneve** [comment]
Excellent! I will try it out right away.

**Pull request merged**